h/api/test/auth_test.py

```diff
@@ -28,3 +31,209 @@ def test_translate_annotation_principals(p_in, p_out):
+
+    assert decode.call_args[1]['leeway'] == 240
+
+
+@user_id_from_bearer_token_fixtures
+def test_userid_from_bearer_token_passes_algorithm_to_decode(decode):
+    request = mock_request(token='Bearer abc123')
+
+    auth.userid_from_bearer_token(request)
+
+    assert decode.call_args[1]['algorithm'] == 'HS256'
+
+
+@user_id_from_bearer_token_fixtures
+def test_userid_from_bearer_token_returns_sub_from_decode(decode):
+    decode.return_value = {'sub': 'acct:test_user@hypothes.is'}
+
+    result = auth.userid_from_bearer_token(mock_request(token='Bearer abc123'))
+
+    assert result == 'acct:test_user@hypothes.is'
+
+
+@user_id_from_bearer_token_fixtures
+def test_userid_from_bearer_token_returns_None_if_no_sub(decode):
+    decode.return_value = {}  # No 'sub' key.
+
+    result = auth.userid_from_bearer_token(mock_request(token='Bearer abc123'))
+
+    assert result is None
+
+
+@user_id_from_bearer_token_fixtures
+def test_userid_from_bearer_token_returns_None_if_decoding_fails(decode):
+    decode.side_effect = jwt.InvalidTokenError
+
+    result = auth.userid_from_bearer_token(mock_request(token='Bearer abc123'))
+
+    assert result is None
+
+
+unauthenticated_userid_from_bearer_token_fixtures = pytest.mark.usefixtures(
+    'decode')
+
+
+@unauthenticated_userid_from_bearer_token_fixtures
+def test_unauthenticated_userid_from_bearer_token_no_Authorization():
+    request = mock_request()
+    request.headers = {}
+
+    result = auth.unauthenticated_userid_from_bearer_token(request)
+
+    assert result is None
+
+
+@unauthenticated_userid_from_bearer_token_fixtures
+def test_unauthenticated_userid_from_bearer_token_calls_decode(decode):
+    request = mock_request(token='Bearer abc123')
+
+    auth.unauthenticated_userid_from_bearer_token(request)
+
+    decode.assert_called_once_with('abc123', verify=False)
+
+
+@unauthenticated_userid_from_bearer_token_fixtures
+def test_unauthenticated_userid_from_bearer_token_returns_userid(decode):
+    request = mock_request(token='Bearer abc123')
+    decode.return_value = {'sub': 'acct:test_user@hypothes.is'}
+
+    result = auth.unauthenticated_userid_from_bearer_token(request)
+
+    assert result == 'acct:test_user@hypothes.is'
+
+
+@unauthenticated_userid_from_bearer_token_fixtures
+def test_unauthenticated_userid_from_bearer_token_when_decode_fails(decode):
+    request = mock_request(token='Bearer abc123')
+    decode.side_effect = jwt.InvalidTokenError
+
+    result = auth.unauthenticated_userid_from_bearer_token(request)
+
+    assert result is None
+
+
+def mock_request(token=None):
+    request = mock.Mock(authenticated_userid='acct:testuser@hypothes.is',
+                        host_url='https://hypothes.is')
+    request.registry.settings = {
+        'h.client_id': 'id',
+        'h.client_secret': 'secret'
+    }
+    if token:
+        request.headers = {'Authorization': token}
+    return request
+
+
+@pytest.fixture
+def encode(request):
```

Comment:
Normally I would do that (normally I prefer to patch top-level names in the module under test, and name the fixture exactly the same as the name being patched in the module) but in this case the tests need to `import jwt` in order to get `jwt.InvalidTokenError` so a fixture named `jwt` would clash. Rather than make up a name like `jwt_fixture` or something I just patched `encode` and `decode` instead.